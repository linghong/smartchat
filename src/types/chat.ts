export interface Chat {
  id: number;
  title: string;
}

export type ChatRole = 'system' | 'assistant' | 'user';

export type ChatType = {
  role: ChatRole;
  content: string;
};

export interface ImageFile {
  base64Image: string;
  mimeType: string;
  size: number;
  name: string;
  width?: number;
  height?: number;
}

export interface FileData {
  base64Content: string;
  type: string;
  size: number;
  name: string;
}

export interface Message {
  question: string;
  answer: string;
  model: string;
}

export interface OpenAIChatContentImage {
  type: 'image_url';
  image_url: {
    url: string;
    detail: 'auto' | 'low' | 'high';
  };
}
export interface OpenAIChatContentText {
  type: 'text';
  text: string;
}
