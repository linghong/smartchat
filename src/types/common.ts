export interface OptionType {
  value: string;
  label: string;
  tags?: string[];
  category?: string;
  contextWindow?: number;
  vision?: boolean;
}

export interface ChatOption extends OptionType {
  tags: string[];
}

export type ApiResponse = {
  message: string;
  fileName: string;
  error?: string;
};

export type InputErrors = {
  [key: string]: string | null;
};

export type Input<T> = {
  [key: string]: T;
};

export type InputData = {
  [key: string]: string | number | null;
};

export type UploadData = {
  [key: string]: File | null;
};

export type UploadErrors = {
  [key: string]: string | null;
};
