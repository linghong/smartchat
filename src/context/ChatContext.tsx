'use client';
import React, {
  createContext,
  useContext,
  useState,
  ReactNode,
  Dispatch,
  SetStateAction
} from 'react';

import { Message, FileData } from '@/src/types/chat';
import { OptionType } from '@/src/types/common';
import { initialMessage, defaultModel } from '@/src/utils/initialData';

export interface ChatContextType {
  isNewChat: boolean;
  setIsNewChat: Dispatch<SetStateAction<boolean>>;
  isConfigPanelVisible: boolean;
  setIsConfigPanelVisible: Dispatch<SetStateAction<boolean>>;
  isSearchChat: boolean;
  setIsSearchChat: Dispatch<SetStateAction<boolean>>;
  fileSrcHistory: FileData[][];
  setFileSrcHistory: Dispatch<SetStateAction<FileData[][]>>;
  chatHistory: Message[];
  setChatHistory: Dispatch<SetStateAction<Message[]>>;
  activeChat: OptionType;
  setActiveChat: Dispatch<SetStateAction<OptionType>>;
  chats: OptionType[];
  setChats: Dispatch<SetStateAction<OptionType[]>>;
  selectedModel: OptionType;
  setSelectedModel: Dispatch<SetStateAction<OptionType>>;
}

export const ChatContext = createContext<ChatContextType | undefined>(
  undefined
);

export const ChatProvider: React.FC<{ children: ReactNode }> = ({
  children
}) => {
  const [isNewChat, setIsNewChat] = useState(true);
  const [isConfigPanelVisible, setIsConfigPanelVisible] = useState(false);
  const [isSearchChat, setIsSearchChat] = useState(false);
  const [fileSrcHistory, setFileSrcHistory] = useState<FileData[][]>([[]]);
  const [chatHistory, setChatHistory] = useState<Message[]>([initialMessage]);
  const [activeChat, setActiveChat] = useState<OptionType>({
    value: '0',
    label: '0',
    tags: []
  }); //new chat
  const [chats, setChats] = useState<OptionType[]>([]);
  const [selectedModel, setSelectedModel] = useState<OptionType>(defaultModel);

  return (
    <ChatContext.Provider
      value={{
        isNewChat,
        setIsNewChat,
        isConfigPanelVisible,
        setIsConfigPanelVisible,
        isSearchChat,
        setIsSearchChat,
        fileSrcHistory,
        setFileSrcHistory,
        chatHistory,
        setChatHistory,
        activeChat,
        setActiveChat,
        chats,
        setChats,
        selectedModel,
        setSelectedModel
      }}
    >
      {children}
    </ChatContext.Provider>
  );
};

export const useChatContext = () => {
  const context = useContext(ChatContext);
  if (context === undefined) {
    throw new Error('useChatContext must be used within a ChatProvider');
  }
  return context;
};
