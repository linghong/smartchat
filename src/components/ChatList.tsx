import { FC, useState, useEffect, useCallback } from 'react';

import TagSearch from '@/src/components/TagSearch';
import { useChatContext } from '@/src/context/ChatContext';
import { OptionType } from '@/src/types/common';

const ChatList: FC = () => {
  const { chats } = useChatContext();

  const [filteredChats, setFilteredChats] = useState<OptionType[]>(chats);

  const handleFilterChats = useCallback((newFilteredChats: OptionType[]) => {
    setFilteredChats(newFilteredChats);
  }, []);

  // Update filteredChats when the `chats` prop changes
  useEffect(() => {
    setFilteredChats(chats);
  }, [chats]);

  return (
    <div>
      <TagSearch onFilterChats={handleFilterChats} />
      <div className="flex-grow overflow-y-auto bg-white border-2 border-stone-200">
        {filteredChats.length > 0 ? (
          <div className="p-4">
            <h3 className="font-bold mb-2">Filtered Chats:</h3>
            {filteredChats.map(chat => (
              <div key={chat.value} className="mb-2 p-2 bg-gray-100 rounded">
                <p>{chat.label}</p>
                {chat.tags && (
                  <div className="mt-1">
                    {chat.tags.map(tag => (
                      <span
                        key={tag}
                        className="mr-2 px-2 py-1 bg-blue-50 text-blue-800 rounded-full text-xs"
                      >
                        {tag}
                      </span>
                    ))}
                  </div>
                )}
              </div>
            ))}
          </div>
        ) : (
          <div className="p-4 text-center text-gray-500">
            No chats found matching the selected tags.
          </div>
        )}
      </div>
    </div>
  );
};

export default ChatList;
