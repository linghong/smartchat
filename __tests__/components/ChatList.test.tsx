// ChatList.test.tsx
import React from 'react';
import { screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';

import ChatList from '@/src/components/ChatList';
import { renderWithContext } from '@/__tests__/test_utils/context';
import { OptionType } from '@/src/types/common';

// Mock the TagSearch component
jest.mock('@/src/components/TagSearch', () => ({
  __esModule: true,
  default: ({
    onFilterChats
  }: {
    onFilterChats: (chats: OptionType[]) => void;
  }) => (
    <button
      data-testid="mock-tag-search"
      onClick={() =>
        onFilterChats([
          { value: '1', label: 'Chat 1', tags: ['tag1'] },
          { value: '2', label: 'Chat 2', tags: ['tag2'] }
        ])
      }
    >
      Mock TagSearch
    </button>
  )
}));

describe('ChatList Component', () => {
  const mockChats: OptionType[] = [
    { value: '1', label: 'Chat 1', tags: ['tag1', 'tag2'] },
    { value: '2', label: 'Chat 2', tags: ['tag2'] },
    { value: '3', label: 'Chat 3', tags: ['tag3'] }
  ];
  const emptyChats: OptionType[] = [];

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render the TagSearch component', () => {
    renderWithContext(<ChatList />, { chats: mockChats });

    expect(screen.getByTestId('mock-tag-search')).toBeInTheDocument();
    expect(screen.getByText('Mock TagSearch')).toBeInTheDocument();
  });

  it('should render the list of chats when filteredChats has items', () => {
    renderWithContext(<ChatList />, { chats: mockChats });

    expect(screen.getByText('Filtered Chats:')).toBeInTheDocument();

    mockChats.forEach(chat => {
      const chatElement = screen
        .getByText(chat.label)
        .closest('div.bg-gray-100');
      expect(chatElement).toBeInTheDocument();
      expect(screen.getByText(chat.label)).toBeInTheDocument();
      if (chat.tags) {
        chat.tags.forEach(tag => {
          expect(screen.getAllByText(tag).length).toBeGreaterThan(0);
        });
      }
    });
  });

  it('should display "No chats found..." when filteredChats is empty', () => {
    renderWithContext(<ChatList />, { chats: emptyChats });

    expect(
      screen.getByText('No chats found matching the selected tags.')
    ).toBeInTheDocument();
  });

  it('should update the displayed chats when TagSearch filters chats', async () => {
    renderWithContext(<ChatList />, { chats: mockChats });

    expect(screen.getByText('Chat 1')).toBeInTheDocument();
    expect(screen.getByText('Chat 2')).toBeInTheDocument();
    expect(screen.getByText('Chat 3')).toBeInTheDocument();

    fireEvent.click(screen.getByTestId('mock-tag-search'));

    await waitFor(() => {
      expect(screen.getByText('Chat 1')).toBeInTheDocument();
      expect(screen.getByText('Chat 2')).toBeInTheDocument();
      expect(screen.queryByText('Chat 3')).not.toBeInTheDocument();
    });
  });

  it('should match the snapshot', () => {
    const { asFragment } = renderWithContext(<ChatList />, {
      chats: mockChats
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
