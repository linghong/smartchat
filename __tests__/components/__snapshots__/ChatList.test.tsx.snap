// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatList Component should match the snapshot 1`] = `
<DocumentFragment>
  <div>
    <button
      data-testid="mock-tag-search"
    >
      Mock TagSearch
    </button>
    <div
      class="flex-grow overflow-y-auto bg-white border-2 border-stone-200"
    >
      <div
        class="p-4"
      >
        <h3
          class="font-bold mb-2"
        >
          Filtered Chats:
        </h3>
        <div
          class="mb-2 p-2 bg-gray-100 rounded"
        >
          <p>
            Chat 1
          </p>
          <div
            class="mt-1"
          >
            <span
              class="mr-2 px-2 py-1 bg-blue-50 text-blue-800 rounded-full text-xs"
            >
              tag1
            </span>
            <span
              class="mr-2 px-2 py-1 bg-blue-50 text-blue-800 rounded-full text-xs"
            >
              tag2
            </span>
          </div>
        </div>
        <div
          class="mb-2 p-2 bg-gray-100 rounded"
        >
          <p>
            Chat 2
          </p>
          <div
            class="mt-1"
          >
            <span
              class="mr-2 px-2 py-1 bg-blue-50 text-blue-800 rounded-full text-xs"
            >
              tag2
            </span>
          </div>
        </div>
        <div
          class="mb-2 p-2 bg-gray-100 rounded"
        >
          <p>
            Chat 3
          </p>
          <div
            class="mt-1"
          >
            <span
              class="mr-2 px-2 py-1 bg-blue-50 text-blue-800 rounded-full text-xs"
            >
              tag3
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
